### PR TITLE
Allow copying multiple commits

### LINKED ISSUE #1244: Ranged copy hash
**Is your feature request related to a problem? Please describe.**
Cherry picking is a daily occurence for my work, and it would usually involve a lot of commits needing to be picked, often 1, but often 10-20 commits as well.

**Describe the solution you'd like**
It would be nice if we can do a commit selection and have the copy hash shortcut copy the range.

**Describe alternatives you've considered**
Yes, I could just do a copy and paste of every single commits I want picked, however that's not fast and convenient.

**Additional context**
None.

### PR DESCRIPTION
**Pull request by remique**
This Pull Request fixes #1244.

It changes the following:
- allows copying multiple commits by marking them 

As of now, the functionality is as follows:
- No commits marked: copies selected line with short commit id
- One commit marked: copies ~~long~~ short commit id
- More than one commit marked **consecutively**: copies `c1^..c8` (short hashes)
- More than one commit marked **randomly**: copies `c1 c2 c4 c5 c8` (short hashes)

I followed the checklist:
- [ ] I added unittests
- [x] I ran `make check` without errors
- [x] I tested the overall application
- [x] I added an appropriate item to the changelog

### CODE REVIEW

**Comment by remique** [comment]
I have noticed that sometimes the app crashes after I yank the commits. I reckon that it is caused by sorting the indexes and I am looking into it.

**Comment by remique** [comment]
Seems to be working now. I'm not sure if it is the greatest design, but works and does not crash.

**Comment by extrawurst** [comment]
You seem to be saving indices. Did you check if that works still in big repos with the log being loaded in batches?

**Comment by remique** [comment]
I'm only saving indices that have been marked. I tested it with [25k commit](https://github.com/sqlite/sqlite) repo and it works fine.

**Comment by extrawurst** [comment]
can you make the required addition to the changelog?

**Comment by remique** [comment]
Done!

File: src/components/commitlist.rs [hunk 1]
```diff
@@ -35,6 +35,7 @@ pub struct CommitList {
 	count_total: usize,
 	items: ItemBatch,
 	marked: Vec<CommitId>,
+	marked_indexes: Vec<usize>,
```
Comment (extrawurst) [inline src/components/commitlist.rs:38]: please reuse `marked` here. feel free to make it a `Vec<(CommitId,usize)>` for example

Comment (extrawurst) [reply src/components/commitlist.rs:38]: @remique are you still on this?

**Comment by extrawurst** [comment]
putting in draft until the required changes are made

**Comment by remique** [comment]
Hey, I updated the changes and everything seems to be working fine on a big repo. I also ran `make check` locally with no errors. Failed build seems to be a clippy bug though? 

```
error[E0277]: expected a `FnOnce<()>` closure, found `std::string::String`
   --> src/components/commitlist.rs:195:20
    |
195 |                         .map_or_else(String::new(), |le| {
    |                          ----------- ^^^^^^^^^^^^^ expected an `FnOnce<()>` closure, found `std::string::String`
    |                          |
    |                          required by a bound introduced by this call
    |
    = help: the trait `FnOnce<()>` is not implemented for `std::string::String`
    = note: wrap the `std::string::String` in a closure with no arguments: `|| { /* code */ }`
```

File: src/components/commitlist.rs [hunk 2]
```diff
@@ -143,11 +143,83 @@ impl CommitList {
...
+					format!("{}^..{}", f.hash_short, l.hash_short);
+				crate::clipboard::copy_string(&yank)?;
+			};
+		} else {
+			let separate = self
+				.marked_indexes()
+				.iter()
+				.map(|e| {
+					self.items.iter().nth(*e).map_or_else(
+						|| String::from(""),
```
Comment (extrawurst) [inline src/components/commitlist.rs:193]: This is different code than the one you posted from your clippy warning.
the proper way to write this is: `map_or_else(String::new, ...`

**Review by extrawurst — CHANGES_REQUESTED** [review]
fix clippy nightly. see inline comment

File: src/components/commitlist.rs (hunk 2 above)
Comment (remique) [reply src/components/commitlist.rs]: Ah, was using `String::new()` and got an error. This always confuses me. Anyway, fixed that.

**Comment by extrawurst** [comment]
Thank you ❤️

**Pull request merged**